### DISCUSSION TITLE
Implementing Psr\Http\Client\ClientInterface in ReactPHP Browser

### DISCUSSION THREAD

**Discussion by hasanparasteh** [reactphp/reactphp#565] [Ideas]
Hey there,

I wanted to chat about a cool idea for improving ReactPHP Browser, particularly for making it work smoothly with other PHP projects. What I'm suggesting is implementing something called `Psr\Http\Client\ClientInterface` in ReactPHP Browser.

Why is this a big deal? Well, by doing this, ReactPHP Browser would become compatible with a bunch of other PHP projects out there, like [InfluxDB Client](https://github.com/influxdata/influxdb-client-php) PHP, which is pretty popular. This means folks could use ReactPHP Browser more easily in their projects without any headaches.

Here are the key benefits:

1. **Better Compatibility:** Adding `Psr\Http\Client\ClientInterface` means ReactPHP Browser would play nicely with a bunch of other PHP projects that follow the same standards.
  
2. **More Users:** Being compatible with `Psr\Http\Client\ClientInterface` could attract more people to use ReactPHP Browser, especially those who really care about sticking to standards.
  
3. **Community Collaboration:** By doing this, ReactPHP can encourage more teamwork within the PHP community. People who know their way around `Psr\Http\Client\ClientInterface` might be excited to pitch in and make ReactPHP Browser even better.

I'd love to hear your thoughts on this idea and whether it fits into the plans for ReactPHP Browser. And if you need a hand with making it happen, count me in!

Thanks for considering this suggestion. Looking forward to chatting more about it with you and the community.

**Comment by WyriHaximus** [comment]
Hey @hasanparasteh, thank you for this well-written piece. However, we're already intending to do so: https://github.com/orgs/reactphp/discussions/481

**Comment by hasanparasteh** [comment]
Thanks! I didn't noticed that in the v3 discussion

**Comment by SimonFrings** [comment]
@hasanparasteh really like your description, so let me give you a few more insights on this topic 👍️

We talked in our team about this just a few days ago and yes, we  also want to see PSR-15 support in ReactPHP. @clue already looked into an implementation in the past and as we described in https://github.com/orgs/reactphp/discussions/481, we would love to support more PSRs ( like PSR-15, PSR-16, PSR-17, PSR-18) in the future.

We already published our roadmap for HTTP in https://github.com/reactphp/http/issues/517 highlighting the necessary topics for ReactPHP HTTP v3, but we're unsure if adding PSR-15 support needs to be one of them. I think we should focus on the most important steps to upgrade to v3 for now as the rest can be addressed in a follow-up version (v3.1+).

To keep you in the loop, we're currently working on a HTTP v1.10.0 in preparation for a HTTP v3 and will look into the next v3 features afterwards.

If you need PSR-15 support (more) urgently, you can also reach out to us or help us prioritizing by [becoming as sponsor](https://github.com/sponsors/reactphp) ❤️

**Reply by WyriHaximus** [reply]
A few notes on this, most of this is IIRC @hasanparasteh, but in a nutshell:
* We've already started preparing for PSR-16 in cache aligning the method names some time ago
* PSR-18 can easily be added as we don't have conflicting method names, and with v3 we're returning the response instead of the promise resolving to that response. So all we would have to do there is implement the interface with a wrapper method.
* Done PSR-15 in a package somewhere, will make sure it's up to date using fibers some time soon.
* PSR-17 and PSR-18 can already be done through https://github.com/php-http/react-adapter/ today, it uses fibers to be compatible.

**Reply by hasanparasteh** [reply]
What a great work! That's awesome. I found a workaround for my problem using [react-adapter](https://github.com/php-http/react-adapter/) and It works fine for my case. I'm eagerly waiting for `reactphp v3` release!
